File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-typescript/demo/src/report.ts
// Copyright (C) 2026 Antony Stubbs and contributors

/**
 * What one arm achieved, and the two tables the whole demo exists to print.
 *
 * **Throughput only, and no latency anywhere.** The backlog is pre-produced, so the workload is
 * closed-loop: a per-record timing is flattered by however far an arm had fallen behind, and
 * reporting one would make the slower arm look better the further behind it got. Records per
 * second over a fixed backlog is the honest number this shape can produce, so it is the only one.
 */

/** The serial arm's name, spelled the same everywhere. Never bare "core" - that reads as
 * `parallel-consumer-core` (`CONCEPTS.md`). */
export const AK_CORE = "AK core";

/** The sidecar arm's name: this language, over the socket, through this module's client library. */
export const SIDECAR_ARM = "typescript-grpc";

export interface ArmResult {
  readonly arm: string;
  readonly elapsedMs: number;
  readonly processed: number;
}

export function ratePerSecond(result: ArmResult): number {
  return result.elapsedMs > 0 ? (result.processed * 1_000) / result.elapsedMs : 0;
}

/**
 * One replay's table.
 *
 * @param baseline       the AK core row every ratio is against, or undefined when the replay had no
 *                       serial arm to compare with
 * @param acrossReplays  true for the big replay, whose baseline comes from the SMALL replay and is
 *                       therefore not like-for-like. Marked in the column heading and footnoted,
 *                       rather than quietly printed as if it were.
 */
export function table(
  title: string,
  results: readonly ArmResult[],
  baseline: ArmResult | undefined,
  acrossReplays: boolean,
): string {
  const baselineRate = baseline === undefined ? 0 : ratePerSecond(baseline);
  const heading = acrossReplays ? "vs AK core*" : "vs AK core";
  const lines = [
    "",
    title,
    `  ${"arm".padEnd(16)}${"elapsed".padStart(10)}${"msg/s".padStart(14)}${heading.padStart(14)}`,
  ];
  for (const result of results) {
    const ratio = baselineRate === 0 ? "-" : `${(ratePerSecond(result) / baselineRate).toFixed(1)}x`;
    lines.push(
      `  ${result.arm.padEnd(16)}` +
        `${`${(result.elapsedMs / 1_000).toFixed(1)}s`.padStart(10)}` +
        `${Math.trunc(ratePerSecond(result)).toLocaleString("en-US").padStart(14)}` +
        `${ratio.padStart(14)}`,
    );
  }
  if (acrossReplays) {
    lines.push("");
    lines.push("  * against the SMALL replay's AK core arm. Across replays, so not like-for-like.");
  }
  return `${lines.join("\n")}\n`;
}

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-go/demo/report.go
// Copyright (C) 2026 Antony Stubbs and contributors

package main

import (
	"fmt"
	"os"
	"strings"
)

// logf is the demo's whole output channel: plain lines on stdout, no logging framework.
//
// The tables are the product here, and the shared contract fixes their columns and their order, so
// a framework's prefixes and levels would only get between the reader and them. Everything the
// sidecar itself emits goes to stderr, which keeps the two separable by a scripted caller.
func logf(format string, args ...any) {
	fmt.Fprintf(os.Stdout, format+"\n", args...)
}

func errorf(format string, args ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", args...)
}

// report prints one replay's table.
//
// SAME COLUMNS, SAME ORDER, SAME WIDTHS AS EVERY OTHER LANGUAGE'S DEMO, so a reader who has run one
// has run them all - that is the whole point of the shared contract. THROUGHPUT ONLY: the backlog
// is pre-produced, so the workload is closed-loop and per-record timings would be flattered by
// however far an arm fell behind.
//
// baseline is the AK core arm the ratios are against, or nil when there is none to compare with.
// acrossReplays marks the big replay, whose ratios are against the SMALL replay's AK core - which
// is not like-for-like, and says so under the table rather than in a footnote nobody reads.
func report(title string, results []armResult, baseline *armResult, acrossReplays bool) {
	var b strings.Builder
	fmt.Fprintf(&b, "\n\n%s\n", title)

	comparison := "vs AK core"
	if acrossReplays {
		comparison = "vs AK core*"
	}
	fmt.Fprintf(&b, "  %-14s %10s %14s %14s\n", "arm", "elapsed", "msg/s", comparison)

	for _, r := range results {
		ratio := "-"
		if baseline != nil && baseline.ratePerSecond() != 0 {
			ratio = fmt.Sprintf("%.1fx", r.ratePerSecond()/baseline.ratePerSecond())
		}
		fmt.Fprintf(&b, "  %-14s %9.1fs %14s %14s\n",
			r.arm, r.elapsed.Seconds(), thousands(int(r.ratePerSecond())), ratio)
	}
	if acrossReplays {
		b.WriteString("\n  * against the SMALL replay's AK core arm. Across replays, so not like-for-like.\n")
	}
	logf("%s", b.String())
}

// thousands renders a count with comma separators, which Go's fmt has no verb for.
func thousands(n int) string {
	sign := ""
	if n < 0 {
		sign, n = "-", -n
	}
	digits := fmt.Sprintf("%d", n)
	var out strings.Builder
	for i, d := range digits {
		if i > 0 && (len(digits)-i)%3 == 0 {
			out.WriteByte(',')
		}
		out.WriteRune(d)
	}
	return sign + out.String()
}
